walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,98 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    //If there are no events or no one is attending the event, 
+    // all requests are possible if they are within the range of the whole day.
+    if(request.getDuration() <= TimeRange.WHOLE_DAY.duration()){
+      if(events.isEmpty() || request.getAttendees().isEmpty()) {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    } else {
+      // Otherwise no available times.
+      return Arrays.asList();
+    }
+
+    // List of existing bookedEvents for attendees in meeting request.
+    List<Event> bookedEvents = new ArrayList<>();
+
+    // Find all events that attendees in the request are already booked for.
+    for(Event event : events){
+      for (String attendee : request.getAttendees()){
+        if(event.getAttendees().contains(attendee) && !bookedEvents.contains(event)){
+          bookedEvents.add(event);
+        }
+      }
+    }
+    
+    // List of potential free time ranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+
+    if(bookedEvents.isEmpty()){
+      // attendees are not booked for any events, are available for the whole day.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else {
+
+      Event prevEvent = bookedEvents.get(0); 
+      Event currEvent = bookedEvents.get(0);
+      for(int i = 0; i < bookedEvents.size(); i++){ 
```

Comment:
I'd consider pulling out a separate method for this inner loop (for example, passing through the current 'tail' potential time period (initialized to the whole day) along with the next booked event, and returning the 0 to 2 replacement time periods not overlapping with that event.

Reply:
Done